### PR TITLE
ci: revise windows gha workflow

### PR DESCRIPTION
**Pull request by michaelsbradleyjr**
Related to #463 and #1922.

This is just a start, and I don't really know (yet) what I'm doing when making changes to the `.h` files, etc.

For example: `<sys/endian.h>` doesn't exist on Windows + mingw-w64-ucrt-x86_64. However, I found an example where someone was using `<winsock.h>` in its place, so I tried that. My `define htole(x) (bswap32(htonl(x)))` is just a guess, copied from the one in the `else` clause.

`<poll.h>` is also not available, that's the next thing the compiler tripped on when running the `make` step. If it's commented out (just to see what happens), then it fails re: `termios.h` not being available. For that, [`termiWin`](https://github.com/veeso/termiWin) seems like it might be helpful, but it's GPLv3. Maybe there's an altogether better solution.

Hopefully, some other folks who want to see notcurses running on Windows can jump in and help.

Note that the GHA workflow for Windows still needs to be triggered manually from the *Actions* panel.

### CODE REVIEW

**Comment by michaelsbradleyjr** [comment]
For reasons unknown, in this repo's *Actions* panel I'm not currently able to trigger the workflow for this PR branch using the desired worfklow (corresponding to branch `michaelsbradleyjr:ci/revise-windows-gha-workflow`).

However, I am able to do it in my fork's *Actions* panel. Here's the output: https://github.com/michaelsbradleyjr/notcurses/runs/3099663448?check_suite_focus=true.

It gets through `cmake` with flying colors, but `make` fails as described previously.

**Comment by dankamongmen** [comment]
> For reasons unknown, in this repo's _Actions_ panel I'm not currently able to trigger the workflow for this PR branch using the desired worfklow (the revised one on branch `michaelsbradleyjr:ci/revise-windows-gha-workflow`).
> However, I am able to do it in my fork's _Actions_ panel. Here's the output: https://github.com/michaelsbradleyjr/notcurses/runs/3099663448?check_suite_focus=true.

possibly a permissions thing? except you've got rights here, so maybe not.

**Comment by dankamongmen** [comment]
mingw is like a cygwin thing, isn't it? i haven't had to use cygwin in something like twenty years, and very much disliked it when i did. i've been planning on building this up in visual studio -- that's what i've been doing today (mentioned in #463 just now). would we want mingw in addition to VSC++? like, do people use mingw if they don't have a copy of VSC++ (as a microsoft employee, i get my VSC++ for free)? i thought mingw was kinda the poor man's way of building poorly- or non-ported UNIX software on Windows.

all of the above may be completely wrong.

**Comment by michaelsbradleyjr** [comment]
> mingw is like a cygwin thing, isn't it? i haven't had to use cygwin in something like twenty years, and very much disliked it when i did. i've been planning on building this up in visual studio -- that's what i've been doing today (mentioned in #463 just now). would we want mingw in addition to VSC++? like, do people use mingw if they don't have a copy of VSC++ (as a microsoft employee, i get my VSC++ for free)?

There is (or can be) quite a bit of misunderstanding re: msys2, mingw, etc. and what the relationships are to cygwin.

From msys2's website:
> Despite some of these central parts being based on Cygwin, the main focus of MSYS2 is to provide a build environment for **native Windows software** and the Cygwin-using parts are kept at a minimum. MSYS2 provides up-to-date native builds for GCC, mingw-w64, CPython, CMake, Meson, OpenSSL, FFmpeg, Rust, Ruby, just to name a few.

Now that ucrt64 is a thing for msys2, the story is potentially even happier. Some history of that development can be found in https://github.com/msys2/MINGW-packages/issues/6901.

Please also see: https://www.msys2.org/docs/environments/. There is an *MSVCRT vs UCRT* section at the bottom:
> UCRT (Universal C Runtime) is a newer version which is also used by Microsoft Visual Studio by default. It should work and behave as if the code was compiled with MSVC.
> * Better compatibility with MSVC, both at build time and at run time.
> * It only ships by default on Windows 10 and for older versions you have to provide it yourself or depend on the user having it installed.

---

> i thought mingw was kinda the poor man's way of building poorly- or non-ported UNIX software on Windows

Whatever the longer history was (I only became acquainted with msys2 late last year), it seems to offer a rather nice build environment these days. I can and do use it to build standalone native Windows executables using the tools (make, gcc, etc.) that are familiar to me from the Linux/macOS worlds I spend more time in.

I've personally never run Visual Studio, and I found MSYS2 encouraging because it meant I didn't have to learn and use a toolchain completely foreign to me.

I don't want to discourage you from using Visual Studio, though, and I'm 100% sure it's possible to configure GitHub Actions to build with that toolchain. At the same time, it would be great if notcurses ends up being buildable with MSYS2.

**Comment by dankamongmen** [comment]
truly it is a pleasure working with you, @michaelsbradleyjr . your appearance in my life seems evidence of the direct manifestation of the Deity in human affairs.

**Comment by dankamongmen** [comment]
`#define htole(x) (bswap32(htonl(x)))`

yeah this looks fine. it would be good to check the output and verify that it reduces to a no-op on little endian. there's apparently a WSAHtonl in winsock2.h, but i see no advantage from it, and it requires calling WSAStartup (why?!).

**Comment by dankamongmen** [comment]
ok MSYS2 looks like good stuff, excellent

**Comment by dankamongmen** [comment]
> ok MSYS2 looks like good stuff, excellent
ahhhhhhh and this is where mintty comes from

**Comment by WSLUser** [comment]
Visual Studio build chain is the preferred route if you do truly want to have a native Windows build. Msys2 build is NOT native Windows build. It is like you said, the poor man's attempt at porting to Windows. For starters, you'll never be able to optimize your perf to the max extent possible without going through MSBUIlD toolchain.

Make is not native to Windows and doesn't work natively (hence Cygwin/Msys2) but support for Cmake is included with Visual Studio. Cmake should be what you use to build on Windows as it currently beats MSVC in terms of what it can do with optimizations among other things.

**Comment by WSLUser** [comment]
I took at look at termiWin and the description kinda proves my point about performance at least. A library like this needs all the perf optimizations it can get so besides the license being a potential issue, it's just plain not suitable for this project.

**Comment by WSLUser** [comment]
So I'd look at Contour for how you should build it out for Windows using Cmake as Contour uses Cmake to compile on all platforms. https://github.com/contour-terminal/contour/blob/master/.github/workflows/build.yml. Ideally notcurses would be built using Cmake on all platforms as well.

**Comment by michaelsbradleyjr** [comment]
> Msys2 build is NOT native Windows build.

The is factually incorrect in my direct experience — having built *standalone* executables that run correctly when launched from `cmd.exe` after being copied to a Windows machine that has absolutely positively nothing installed from MSYS2 or anything related.

It is also incorrect re: what MSYS aims and claims to do per their website and other docs.

In light of MSYS2 now providing a build toolchain that uses UCRT, what you've written seems to me additionally incorrect.

I will be happy to eat my words, and I don't wish to come across harshly. I just think it's important to be correct about this kind of thing, if for no other reason than to avoid confusing people.

What is your working definition of native Windows build and why would binaries produced with MSYS2 not qualify? cf. *"that run correctly ..."* above for my working definition.

Note: 100% static linking, link-time optimization, and other optimizations with gcc work just fine with MSYS2 UCRT64 and MSYS2 MINGW64. Having switched some of my projects to UCRT64, I did notice performance improvements, e.g. with disk IO, so I'm guessing MSYS2 UCRT64 is the better way to go, if one is going with MSYS2.

> I took at look at termiWin and the description kinda proves my point about performance at least. A library like this needs all the perf optimizations it can get so besides the license being a potential issue, it's just plain not suitable for this project.

I agree re: performance and especially re: the GPLv3 license (for a library like notcurses). Hopefully there's a better way/s.

And I think it would be awesome if it works out so that programs built for Windows can run happily in Windows Terminal, mintty, ConEmu, etc., whether launched from `cmd.exe`, PowerShell, or bash, in any of those terminal emulators. Whether that's a realistic goal I'm not sure.

> Ideally notcurses would be built using Cmake on all platforms as well

That would be awesome. I just think it would also be great if the build logic supported building native executables with Microsoft's toolchain or MSYS2.

**Comment by WSLUser** [comment]
> That would be awesome. I just think it would also be great if the cmake build logic supported building native executables with Microsoft's toolchain or MSYS2

So notcurses does already use cmake but that currently won't be the case for Windows the way this build script is written. As I said, take a look at Contour, they have all platforms listed in their yml using cmake. I don't expect one for one matching as that is a terminal vs a terminal library but should be similar enough. 

> And I think it would be awesome if it works out so that programs built for Windows with notcurses can run happily in Windows Terminal, mintty, ConEmu, etc., whether launched from cmd.exe, PowerShell, or bash, in any of those terminal emulators. Whether that's a realistic goal, I'm not sure.

It realistically can be the case, but we'll need 2 seperate builds for Windows (technically we should only need one but having both would be ideal). One that uses the official MS build chain and one that uses the Cygwin/Msys2 build chain. I don't consider (and neither will any Windows developer) the Msys2/Cygwin build chain to be a native chain that completely supports all the Windows APIs, registry calls, NT syscalls, etc.

 I don't know much about UCRT but if it's that much better than the previous way, then I'm guessing it's found a way to do better translations with user-mode and kernel-mode bits.  I just have firm belief that for notcurses to be considered a true Windows library and get first class support for Windows by MS and terminal projects native to Windows like Windows Terminal (and additionally conhost/openconsole, which the same devs are responsible for) you need the MS build chain. Otherwise, only Mintty would be provide any type of support since it completely bypasses the windows console and does it own things with a bunch a hackery that is best explained by mintty himself.

**Comment by michaelsbradleyjr** [comment]
> Otherwise, only Mintty would be provide any type of support since it completely bypasses the windows console and does it own things with a bunch a hackery that is best explained by mintty himself.

In case you're not aware it's possible, you can with MSYS2 MINGW64 or MSYS2 UCRT64 (and probably the other [MSYS2 environments](https://www.msys2.org/docs/environments/), but I haven't tried) build a TUI with ncurses, launch it under `cmd.exe` in ConEmu or Windows Terminal, etc., and it works. 

ncurses TUIs built with MSYS2 don't work well in Windows Terminal or ConEmu, e.g. colors are messed up and resizing is broken because of a lack of SIGWINCH, but there are ways to overcome that (use `PeekConsoleInputW` and deal with events related to the buffer changing size). But they do work.

Oddly, they don't work at all in mintty (note: my context is MSYS2 not Cygwin). The reason for that isn't clear, but it (and other problems) probably have to do with MSYS2's build of ncurses being [rather dated](https://github.com/msys2/MINGW-packages/issues/3593#issuecomment-842605464) and fallen behind recommendations on Dickey's website. Hopefully it can be caught up, or maybe notcurses for Windows will make it moot.

Main point: you don't have to use mintty. With MSYS2, mintty is mostly used to run build tools, and then the things you build with those tools are run in a different terminal emulator or they launch a GUI or whatever.

This and another comment I'm about to make on #1956 will be my last comments attempting to clear up misinformation re: MSYS2. Going forward I'll just link to my comments in this PR and the forthcoming comment on #1956.

I'm sure you mean well @WSLUser, but what you've been saying seems based either on misinformation or outdated information, or both. Please read up on msys2.org and/or go chat with them on their Discord server.

**Comment by WSLUser** [comment]
> With MSYS2, mintty is mostly used to run build tools

Well as it stands. Mintty provides the most VT support due to very fact it doesn't rely on the Windows console. As of today, it is the most compatible with Xterm as far as terminals on Windows go. And as far as this library goes, which relies heavily on VT graphics protocols, Mintty appears to be the only one able to support that since we're still waiting for the console to gain support or for a pass-through mode to be introduced to ConPTY that would allow any terminal running on Windows to run any VT sequences that may not be supported otherwise such as Sixel. 

So basically at this particular point in time, in order for notcurses to be fully tested out to ensure all it's functionality works, we'd need to test with Mintty. Then we can test the native windows console to see how much of what notcurses can do is able to work.

**Comment by WSLUser** [comment]
> ncurses TUIs built with MSYS2 don't work well in Windows Terminal or ConEmu, e.g. colors are messed up and resizing is broken because of a lack of SIGWINCH, but there are ways to overcome that (use PeekConsoleInputW and deal with events related to the buffer changing size). But they do work.

Fortunately we have a work around for the other terminals - WSL. In WSL, I'm able to run plenty of TUI applications on Windows terminal just fine and it's supported by the console. Only graphic protocol stuff is ignored (intentionally until supported).

**Comment by dankamongmen** [comment]
> I don't want to discourage you from using Visual Studio, though, and I'm 100% sure it's possible to configure GitHub Actions to build with that toolchain. At the same time, it would be great if notcurses ends up being buildable with MSYS2.

absolutely regarding the latter, and from what i've read here, i think that could be a completely functional solution to deliver notcurses on windows. with that said, i've heard so many things about the VS compiler backend over the years, and i'd like to see what kind of assembly it generates for notcurses. i'm not at all satisfied with the level of vectorization we're getting out of either gcc or clang, and regularly consider moving to intrinsics in certain core routines. i'd definitely like to avoid that if i can, and if VS can autovectorize better, i'd rather try rearranging things in the hope that gcc/clang figure things out. if MSVS is at the same level, i'll likely move to intrinsics and spell everything out.

**Comment by dankamongmen** [comment]
@michaelsbradleyjr if we're going to be able to see the make output (assuming we can get the action enabled in here), i.e. i can develop iteratively against `termios.h` missing etc., i think i'd like you to go ahead and merge this so that i can start on that effort. i've got a windows branch i'm sitting on which i think might fix up most of the issues you're seing.

**Comment by michaelsbradleyjr** [comment]
Will be finalized and merged in the next hour or so, and will run on all commits to `master` and PRs.

**Comment by dankamongmen** [comment]
i am wetting myself in anticipation

**Comment by michaelsbradleyjr** [comment]
Merging.
